image/descriptor.go

```diff
@@ -50,17 +51,26 @@ func (d *descriptor) hash() string {
 
 func listReferences(w walker) (map[string]*descriptor, error) {
 	refs := make(map[string]*descriptor)
+	var index v1.ImageIndex
 
 	if err := w.walk(func(path string, info os.FileInfo, r io.Reader) error {
-		if info.IsDir() || !strings.HasPrefix(path, "refs") {
+		if info.IsDir() || filepath.Clean(path) != "index.json" {
 			return nil
 		}
 
-		var d descriptor
-		if err := json.NewDecoder(r).Decode(&d); err != nil {
+		if err := json.NewDecoder(r).Decode(&index); err != nil {
 			return err
 		}
-		refs[info.Name()] = &d
+
+		var d descriptor
+		for i := 0; i < len(index.Manifests); i++ {
+			if index.Manifests[i].Descriptor.Annotations["org.opencontainers.ref.name"] != "" {
+				d.MediaType = index.Manifests[i].Descriptor.MediaType
+				d.Digest = string(index.Manifests[i].Descriptor.Digest)
+				d.Size = index.Manifests[i].Descriptor.Size
+				refs[index.Manifests[i].Descriptor.Annotations["org.opencontainers.ref.name"]] = &d
```

Comment:
This will completely flatten the set of references (silently ignoring duplicates). I've spent quite a while trying to come up with a sane API for `umoci` and I think I've decided the only way to implement this is to remove the concept of a reference in the lower-level libraries and have it as a higher-level construct.

In `umoci` this split is quite easy to do `oci/cas` vs `oci/casext` which adds implementation-agnostic extensions to `oci/cas`. Maybe that's something you can do so you can remove the old concepts of a reference being a simple mapping.